### ISSUE TITLE
Some suggestions

### ISSUE THREAD

**Issue by amariq** [Jartreg/typed-events#2]
### Nice EventEmitter, thanks!

Would you consider to add "context" argument when attaching and removing listeners to avoid using `function#bind(this)`?

Also I think there could be unexpected behavior at `#emit(...)` during execution of handlers if you start to remove other handlers of undergoing event.

**Comment by Jartreg** [comment]
Thank you for your suggestions!

Adding an extra "context" argument is a convenient idea, but I think it's out of scope for this package. I'm trying to keep it as small as possible and adding a custom handling of contexts would add more complexities. To be able to add the same listener multiple times with different contexts, I would need to save both the listener function and the contexts and compare them when removing a listener. It seems to me like that should rather be a responsibility of the caller.
I am however planning to make `#on(...)` and `#once(...)` return a callback which can also be used to remove the listener. When using classes I can also recommend using [bind-decorator](https://www.npmjs.com/package/bind-decorator), which automatically binds class methods marked with `@bind`.

Concerning the unexpected behaviour, thanks for pointing that out!

**Comment by amariq** [comment]
> Adding an extra "context" argument is a convenient idea, but I think it's out of scope for this package. I'm trying to keep it as small as possible and adding a custom handling of contexts would add more complexities

Well, I agree that smaller is better, but I don't think such feature will add such a big overhead to be afraid of. 🙂
In my opinion, "context" support (as optional argument) could make a lot easier working with collections and/or when attaching lots of listeners to different objects -> user will be able, for example, to detach all or specific listeners by just providing "context" object or original listener function.
Of course, using  `@bind` and replacing original functions with bound ones will help, but if you try to avoid creating more functions by using `@bind` then such task just becomes impossible. Also it is kind of prone to error when it comes to adding more listeners to existing code since you not only should remember to bind method and attach it, but also remember to detach it, whereas with "context" support you basically don' need to care about it as long as you can provide "context" parameter when attaching a listener and just use single code line to detach all listeners by provided "context". In short, I think "context" argument is less code and less memory usage (especially when operating

**Comment by amariq** [comment]
> Adding an extra "context" argument is a convenient idea, but I think it's out of scope for this package. I'm trying to keep it as small as possible and adding a custom handling of contexts would add more complexities

Well, I agree that smaller is better, but I don't think such feature will add big enough overhead to be afraid of it. 🙂

In my opinion, "context" support (as optional argument) could make a lot easier working with collections and/or when attaching lots of listeners to different objects -> you will be able, for example, to detach all or specific listeners by just providing "context" object or original listener function.
Of course, using  `@bind` and replacing original functions with bound ones will help, but if you try to avoid creating more functions by using `@bind` then such task just becomes impossible. Also it is kind of prone to error when it comes to adding more listeners to existing code since you not only should remember to bind method and attach it, but also remember to detach it, whereas with "context" argument you basically don' need to worry about it as long as you can provide "context" parameter when attaching a listener and just use single code line to detach all listeners by provided "context".

In short, I think "context" argument will benefit in less code and more convenience for end-user and less memory usage (especially when working with collections of objects). 😊


> When using classes I can also recommend using bind-decorator, which automatically binds class methods marked with @bind.

Thank you, looks handy! 👍

**Comment by Jartreg** [comment]
I'm closing this issue since I still don't think this is a needed feature
